application/src/main/java/org/togetherjava/tjbot/commands/system/CommandSystem.java

```diff
@@ -0,0 +1,166 @@
+package org.togetherjava.tjbot.commands.system;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.events.ReadyEvent;
+import net.dv8tion.jda.api.events.interaction.ButtonClickEvent;
+import net.dv8tion.jda.api.events.interaction.SelectionMenuEvent;
+import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+import net.dv8tion.jda.api.interactions.commands.Command;
+import net.dv8tion.jda.api.interactions.components.ComponentInteraction;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.togetherjava.tjbot.commands.CommandRegistry;
+import org.togetherjava.tjbot.commands.SlashCommand;
+import org.togetherjava.tjbot.db.Database;
+import org.togetherjava.tjbot.util.TriConsumer;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * The command system is the core of command handling in this application.
+ * <p>
+ * It knows and manages all commands, registers them towards Discord and is the entry point of all
+ * events. It forwards events to their corresponding commands and does the heavy lifting on all sort
+ * of event parsing.
+ * <p>
+ * <p>
+ * Commands are made available via {@link CommandRegistry}, then the system has to be added to JDA
+ * as an event listener, using {@link net.dv8tion.jda.api.JDA#addEventListener(Object...)}.
+ * Afterwards, the system is ready and will correctly forward events to all commands.
+ */
+public final class CommandSystem extends ListenerAdapter implements SlashCommandProvider {
+    private static final Logger logger = LoggerFactory.getLogger(CommandSystem.class);
+
+    private final Map<String, SlashCommand> nameToSlashCommands;
+    private static final String RELOAD_COMMAND = "reload";
+
+    /**
+     * Creates a new command system which uses the given database to allow commands to persist data.
+     * <p>
+     * Commands are fetched from {@link CommandRegistry}.
+     *
+     * @param database the database that commands may use to persist data
+     */
+    @SuppressWarnings("ThisEscapedInObjectConstruction")
+    public CommandSystem(@NotNull Database database) {
+        nameToSlashCommands = CommandRegistry.createSlashCommands(database)
+            .stream()
+            .collect(Collectors.toMap(SlashCommand::getName, Function.identity()));
+
+        if (nameToSlashCommands.containsKey(RELOAD_COMMAND)) {
+            throw new IllegalStateException(
+                    "The 'reload' command is a special reserved command that must not be used by other commands");
+        }
+        nameToSlashCommands.put(RELOAD_COMMAND, new ReloadCommand(this));
+
+        if (logger.isInfoEnabled()) {
```

Comment:
Shouldn't `logger.info` check that for you anyways?

Reply:
Yes but then you have to compute the arguments of the call first. Which, in this case, is potentially expensive if the list is long. So I opted to guard it.